dask/dataframe/tests/test_dataframe.py

```diff
@@ -5116,3 +5116,40 @@ def reduce_fn(series):
         .compute()[0]
     )
     assert result == {"x": 14, "y": 64}
+
+
+def test_set_index_with_dask_dt_index():
+    ## was failing, now works
+    # https://github.com/dask/dask/issues/8359
+
+    _values = {
+        "x": [1, 2, 3, 4] * 3,
+        "y": [10, 20, 30] * 4,
+        "name": ["Alice", "Bob"] * 6,
+    }
+    _dates = pd.date_range(start="2022-02-22", freq="16h", periods=12)
+    _df = pd.DataFrame(_values, index=_dates)
+    one_day = pd.Timedelta(days=1)
+
+    demo_ts = dd.from_pandas(_df, npartitions=3)
+    no_dates = dd.from_pandas(pd.DataFrame(_values), npartitions=3)
+
+    # specify a different date index entirely
+    day_index = demo_ts.index.dt.floor("D")
+    day_df = demo_ts.set_index(day_index)
+    expected = dd.from_pandas(
+        pd.DataFrame(_values, index=_dates.floor("D")), npartitions=3
+    )
+    assert_eq(day_df, expected)
```

Comment:
Beyond this, it's also important to check that this operation was cheap. At a minimum, `assert day_df.divisions == day_index.divisions` (divisions should be preserved). You might also want to use `hlg_layer_topological` or the like to confirm that there's only been a `map_partitions` operation added to the graph, not shuffling. Ideally you'd also confirm that `_calculate_divisions` was not called, but that's annoying to do cleanly (we prefer to avoid mocks/patching).

Reply:
Thanks. This makes a lot of sense.